beacon_node/store/src/hot_cold_store.rs

```diff
@@ -2797,77 +2826,55 @@ impl<E: EthSpec, Hot: ItemStore<E>, Cold: ItemStore<E>> HotColdDB<E, Hot, Cold>
             "data_availability_boundary" => data_availability_boundary,
         );
 
-        let mut ops = vec![];
-        let mut last_pruned_block_root = None;
+        let mut guard = self.block_cache.lock();
 
-        for res in self.forwards_block_roots_iterator_until(oldest_blob_slot, end_slot, || {
-            let (_, split_state) = self
-                .get_advanced_hot_state(split.block_root, split.slot, split.state_root)?
-                .ok_or(HotColdDBError::MissingSplitState(
-                    split.state_root,
-                    split.slot,
-                ))?;
+        let remove_blob_if = |blobs_bytes: &[u8]| {
+            let blobs = BlobSidecarList::from_ssz_bytes(blobs_bytes)?;
+            let Some(blob): Option<&Arc<BlobSidecar<E>>> = blobs.first() else {
+                return Ok(false);
+            };
 
-            Ok((split_state, split.block_root))
-        })? {
-            let (block_root, slot) = match res {
-                Ok(tuple) => tuple,
-                Err(e) => {
-                    warn!(
-                        self.log,
-                        "Stopping blob pruning early";
-                        "error" => ?e,
-                    );
-                    break;
-                }
+            if blob.slot() <= end_slot {
+                // Delete from the cache
+                guard.delete_blobs(&blob.block_root());
+                // Delete from the on-disk db
+                return Ok(true);
             };
+            Ok(false)
+        };
 
-            if Some(block_root) != last_pruned_block_root {
-                if self
-                    .spec
-                    .is_peer_das_enabled_for_epoch(slot.epoch(E::slots_per_epoch()))
-                {
-                    // data columns
-                    let indices = self.get_data_column_keys(block_root)?;
-                    if !indices.is_empty() {
-                        trace!(
-                            self.log,
-                            "Pruning data columns of block";
-                            "slot" => slot,
-                            "block_root" => ?block_root,
-                        );
-                        last_pruned_block_root = Some(block_root);
-                        ops.push(StoreOp::DeleteDataColumns(block_root, indices));
-                    }
-                } else if self.blobs_exist(&block_root)? {
-                    trace!(
-                        self.log,
-                        "Pruning blobs of block";
-                        "slot" => slot,
-                        "block_root" => ?block_root,
-                    );
-                    last_pruned_block_root = Some(block_root);
-                    ops.push(StoreOp::DeleteBlobs(block_root));
-                }
-            }
+        self.blobs_db
+            .delete_while(DBColumn::BeaconBlob, remove_blob_if)?;
```

Comment:
This should not be a `delete_while`, because the blobs are in an arbitrary order (by block root), so we don't want to stop deleting as soon as we hit a blob that should not be pruned.

I think we need to iterate the whole column, collect all the block roots of blobs to delete, and then delete all of them using `KeyValueStoreOp::Delete` or similar.

Reply:
Nice! My bad, I didn't actually look at the impl!